homeassistant/components/here_travel_time/config_flow.py

```diff
@@ -130,36 +139,120 @@ async def async_step_user(
     ) -> FlowResult:
         """Handle the initial step."""
         errors = {}
-        options: dict[str, Any] = {}
         user_input = user_input or {}
         if user_input:
             if self.source == config_entries.SOURCE_IMPORT:
-                user_input, options = self._transform_import_input(user_input)
-                # We need to prevent duplicate imports
-                if any(
-                    is_dupe_import(entry, user_input, options)
-                    for entry in self.hass.config_entries.async_entries(DOMAIN)
-                    if entry.source == config_entries.SOURCE_IMPORT
-                ):
-                    return self.async_abort(reason="already_configured")
+                return await self.async_import(user_input)
             try:
-                await self.hass.async_add_executor_job(validate_input, user_input)
+                await self.hass.async_add_executor_job(
+                    validate_api_key, user_input[CONF_API_KEY]
+                )
+                self._config = user_input
+                return self.async_show_menu(
+                    step_id="user",
+                    menu_options=["origin_coordinates", "origin_entity"],
+                )
             except InvalidCredentialsError:
                 errors["base"] = "invalid_auth"
-            except Exception:  # pylint: disable=broad-except
-                _LOGGER.exception("Unexpected exception")
+            except Exception as error:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception: %s", error)
                 errors["base"] = "unknown"
-            else:
-                return self.async_create_entry(
-                    title=user_input["name"], data=user_input, options=options
-                )
-
         return self.async_show_form(
             step_id="user", data_schema=get_user_step_schema(user_input), errors=errors
         )
 
+    async def async_step_origin_coordinates(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Configure origin by using gps coordinates."""
+        if user_input is not None:
+            assert self._config is not None
+            self._config[
+                CONF_ORIGIN
+            ] = f"{user_input[CONF_ORIGIN]['latitude']},{user_input[CONF_ORIGIN]['longitude']}"
```

Comment:
Storing `user_input[CONF_ORIGIN]` here so the lat/long are stored separately, then joined by the integration when loaded, would make this work similar to how the entity format is stored (as separate entity attributes). This can let some of the logic get removed for validation since it can always join the two numbers together and know they are in the right format (and also know the format returned by the entity is always correct, assuming its not None.

Reply:
Where can we remove the validation logic?
At first I thought [here](https://github.com/home-assistant/core/pull/69212/files#diff-6f8ad64aba36a02160f2a8fb5451201a7838b5d8f427e952aaf0dcc9ca65c49eR236) but it then occurred to me that the check would still have to be there because we have to validate the result of `find_coordinates`.

Do you see another spot where splitting in lat/lon would save logic?

Reply:
Seems like the link to the line number didn't work. I meant this:
![image](https://user-images.githubusercontent.com/1481961/166234712-4e4f4481-c125-4c7c-ba70-88055edd57a2.png)

Reply:
I see this is also the same format as the configuration.yaml so it seems like it makes import simpler too. You might consider just reusing the same existing key names since they seem reasonable? thought he "destination": {"latitude":X, "longitude": Y} is also fine.

Generally, i think the existing config flow format is simpler (e.g. CONF_ARRIVAL /CONF_DEPARTURE seems simpler than vs CONF_TIME_TYPE which is deconstructed back into an arrival/departure anyway)

Reply:
The use of CONF_TIME_TYPE is a workaround for the constraint that you can provide either an arrival time or an departure time but not both. I copied it from Google and Waze. Can you think of a better way?

Reply:
I was thinking how it is in configuration yaml already seems fine.  I agree its nice to have it be very strict, but you can also apply validation on the states even if they are separate fields.  Since they are all split out as separate variables in the integration anyway, and both passed to the routing library, and separate already in the configuration yaml it seems simplest to keep it that way in the configuration yaml.  As is, there are conversions batch and forth between a "type" and a single "field" vs two separate fields, so just keeping it as two separate fields seems most straight forward.

I went and looked at the google travel time sensor and it seems like its keeping them separate like I was imagining was simpler here too https://github.com/home-assistant/core/blob/355445db2d5c94dff91f3ea9799ec0ebe2581e4f/homeassistant/components/google_travel_time/sensor.py#L192
```
        dtime = options_copy.get(CONF_DEPARTURE_TIME)
        atime = options_copy.get(CONF_ARRIVAL_TIME)
```

Reply:
I now realize that this is a restriction from back when there wasn't the menu_step. It was the only way to enforce a valid configuration using the UI.

Reply:
Will add this tomorrow

Reply:
I split up the lat/lon as suggested and addressed the arrival/departure time by using menus and TImeSelectors